Parsing improvements for #31

SL-Gundam opened the pull request:
This fully fixes Variant5 #31 

Also added an option for ConverterExtra to disable adding the CSS class after a tag. I do need ConverterExtra for table support so this CSS class was an annoying addition for me

Let me know what you think

SL-Gundam committed:
Handle numeric argument values without quotes

SL-Gundam committed:
Adjust fixInlineElementSpacing to not trigger for emptyTags

SL-Gundam committed:
Allow to disable adding the CSS class after the tag

SL-Gundam committed:
Adjusted test case for this commit d0da9e0fae21d160bef3e6c16dbfd5d0fe7049b9

SL-Gundam commented:
Here an exaplanation behind the commits that fix some handling issues
cee262f9211f6bdd3e353e5f51462defba4ae8ae - Fixes the handling of the font tag in Variant5
d0da9e0fae21d160bef3e6c16dbfd5d0fe7049b9 - Fixes extra line endings that should not be present for various examples in #31

tzi commented:
I'm pretty OK to merge the following feature *if* you add associated test cases: 
 - Adjust fixInlineElementSpacing to not trigger for emptyTags
 - Allow to disable adding the CSS class after the tag

I'm not sure about the numeric arguments, because in the HTML spec you can have string attribute value without quotes.
So, It seems to be more a bug about the attribute value parsing than to allow numeric attribute value.

What do you think?

Cheers,
Thomas.

tzi commented:
I separated 2 of your commits about *inline space fixing* in #33, because it's valuable! 🎉

SL-Gundam commented:
Regarding the test cases: I am planning on adding those but before i spend time on that i wanted to at least have your opinion about the work in progress.
I'm still working on the other examples in #31 so it might be i'll work on them first and add the test cases after that.

Regarding attribute values:
Variant2 will have an unquoted string attribute value so your right that the current fix is not enough

SL-Gundam committed:
Fix URL difference on ending slash presence

SL-Gundam committed:
Handle unquoted attribute values

SL-Gundam committed:
Escape all * and _ instead of just 1 or 2

SL-Gundam commented:
Variant 4 and 5 (#31) are now fixed

6f40bccfe86981f106c3407439fd732c45f43bb9 - Ignore the last / for url and title comparison (Variant4)
56f44248ab9fa1ed663d7786597f4a5272660740 - Changed numeric attribute check cee262f9211f6bdd3e353e5f51462defba4ae8ae to quoted and unquoted
ea771c3774deb13ee362ce08a55b76166c600f80 - Variant4 had "*** text here ***". It was escaped incorrectly. This fixes that

I will look into the failing unit tests later since its pretty late now

SL-Gundam committed:
Cleanup redundant spaces

SL-Gundam committed:
One str_replace instead of three

SL-Gundam commented:
e0241971af582ba3cd6ec0d8c6b82891ba9251b9 - Cleans up redundant spaces using rtrim. Since a lot of tests are expecting these spaces, they are currently failing

SL-Gundam committed:
Adjust testcase for ea771c3774deb13ee362ce08a55b76166c600f80

SL-Gundam committed:
Add back the final rtrim that was removed in e0241971af582ba3cd6ec0d8c6b82891ba9251b9

SL-Gundam committed:
Correct test case for e0241971af582ba3cd6ec0d8c6b82891ba9251b9

SL-Gundam committed:
Merge branch 'master' into Parsing_Improvements

SL-Gundam commented:
That should fix all Test cases for the previously proposed changes

SL-Gundam committed:
Change all html EOLs to line feeds

SL-Gundam committed:
flushLinebreaks added before handling text

SL-Gundam committed:
Decreased the number of lineBreaks after blockelements

SL-Gundam committed:
Added ltrim for html content after closing p tag

SL-Gundam commented:
Some improvement for handling Variant2 in #31 

7a0f6a1f92eddf8ae4bfc00a2bf231ac2fb8fc81 - Markdownify only works properly with line feed EOLs so any incoming HTML is now converted
e4a1f9b288ac7d047dab962346da3c5ef4b89265 - After a closing p tag the line breaks need to be flushed before text is outputted
a72f108370e3e7e316c7d2ee8ebb450c1dd2185d - Variant2 uses div tags around various lines. Div tags only take up one empty line so decreased this here to improve the layout of the from, to, sent and subject lines.
75cf897f806bf0e33d977fa51ec232c1094ec392 - After a closing p tag a ltrim was not performed. This caused any text after a closing p tag to have a space in front of it. Since a ltrim was performed after br tags, i added these as well for p tags

SL-Gundam committed:
Ignore Office namespace o:p tags

SL-Gundam commented:
552ba1b22f14b87b1cc9e492944b0426bf7c6d69 - Ignores Office namespace tag <o:p xmlns:o="#unknown"></o:p>.
I could not find any documentation on other possible office namespace tags. So further tags will have to be added when we encounter them

SL-Gundam committed:
Add function getescapeInText

SL-Gundam committed:
Fix header markdown escaping

SL-Gundam committed:
Add escaping for = markdown headers

SL-Gundam committed:
Add proper amount of slashes for escape regex's

SL-Gundam commented:
0bea029f0e915f955ec943a674e6584666bed9c6 - I had an issue where plaintext content would be parsed for markdown with undesired results. This addition allows me to pull the regular expressions out of Markdownify and apply only the escaping on my content.
475af0004e47baeef2514dbae8027de1dd354928 - The Header escaping did not work as desired if the header wasn't on the first line of the text. This modifies that so that it works properly
dc77382d05826526b31dc289b402a65de361a827 - This escapes any header markdown with =
27131acbcbe0282c62f4784eb64e3070c296f2fc - Trivial change but escapes the regular expressions properly

SL-Gundam committed:
Correction incase the last attribute is unquoted

SL-Gundam committed:
Replace &nbsp; with normal space

SL-Gundam committed:
Add more character to escapeInText

SL-Gundam commented:
Some more improvements

f6a32907bc4a4d693505c4508b8c0ae0ecfbffc5 - If an unquoted attribute ended the tag, the tag closing character would not be properly detected without this
9efd59c06995602d19a76695c2ea89801bec9980 - Some trivial PSR coding standards and add &nbsp; conversion to normal space
fc34c2757137d2ca13447c0ab49a734dcc807a94 - More (|) character is now escaped

nlisgo commented on src/Converter.php:
```diff
@@ -302,6 +304,16 @@ public function setKeepHTML($keepHTML)
+     *
+     * @return array escapeInText
+     */
+    public function getescapeInText()
```
Nitpick: should this be `getEscapeInText`?

SL-Gundam replied:
Most of the variables use a construction where the first word is not capped like here https://github.com/Elephant418/Markdownify/blob/master/src/Converter.php#L169 and here https://github.com/Elephant418/Markdownify/blob/master/src/Converter.php#L185

I kept the function name to adhere to the same name as the variable being retrieved. @tzi so far has not commented on this. But i think he will go over the code when I've finished all of my "improvements" and added the associated test cases per his request

So the short question is. Should the name of the function be 100% exactly the same as the variable being retrieved? or are there guidelines which decide the function name?

SL-Gundam commented:
@tzi How to best handle the following code if encountered in html

> <![if !supportLists]>
> <![endif]>

It's comparable to the <o:p></o:p> but that one at least started and ended with the same tag
Would it be best to make it part of $emptyTags? like
> '![if'
> '![endif]'

SL-Gundam committed:
Merge branch 'master' of github.com:Elephant418/Markdownify into Parsing_Improvements

SL-Gundam committed:
Merge branch 'master' of github.com:Elephant418/Markdownify into Parsing_Improvements

SL-Gundam committed:
Allow numbers in xmlns attributes names

SL-Gundam commented:
xmlns can have numbers in them apparently (Outlook 2010)

`xmlns:udcp2p="http://schemas.microsoft.com/data/udc/parttopart" `

SL-Gundam committed:
Fix empty table tag

SL-Gundam committed:
Correct indentation

SL-Gundam commented:
I had a situation where an html email contained an empty table tag (microsoft onedrive email). This makes sure that this does not result in an error: https://github.com/Elephant418/Markdownify/pull/32/commits/b9b3f416bce84c81046282ef3011413312d20b9d

tzi committed:
Fix PHP8.3 support

SL-Gundam committed:
Merge pull request #1 from Elephant418/fix-php8.3

Fix PHP8.3 support

tzi commented:
There is some parts that are too risky to merge:

 - The broad escapeInText rewrite in src/Converter.php
    It changes core escaping behavior for *, _, |, backticks, headings, and links all at once. Too invasive.
  - The cleanup change that trims every output line
    That can subtly alter Markdown semantics, especially around hard breaks and indentation-sensitive output.
  - The change from setLineBreaks(2) to setLineBreaks(1) in handleTagToText()
    That is behavior-changing and likely to create regressions.
  - The “autolink if URLs differ only by trailing slash” change
    It changes link rendering semantics and should be discussed separately.
  - Test expectation rewrites for existing <br> behavior and blockquote blank lines
    Those are not simple modernization changes; they redefine output format.

tzi commented:
I will keep 3 features from your proposal:

  - unquoted attribute parsing
  - optional `o:p` handling
  - optionally the `setAddCssClass(false)` feature
    
Thanks for your help.

Cheers,
Thomas.
